llvm/lib/Transforms/Utils/LoopMetadata.cpp

```diff
@@ -0,0 +1,235 @@
+
+#include "llvm/Transforms/Utils/LoopMetadata.h"
+#include "llvm/Analysis/AssumptionCache.h"
+#include "llvm/IR/Dominators.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/Transforms/Utils/LoopUtils.h"
+
+#define DEBUG_TYPE "loop-metadata"
+
+using namespace llvm;
+
+/// extract the step size by which the loop IV changes.
+int getSCEVStepSize(const SCEV *S);
+
+PreservedAnalyses LoopMetadata::run(Loop &L, LoopAnalysisManager &AM,
+                                    LoopStandardAnalysisResults &AR,
+                                    LPMUpdater &U) {
+  SE = &AR.SE;
+  AC = &AR.AC;
+  DT = &AR.DT;
+
+  assignLoopMetadata(L);
+
+  return PreservedAnalyses::all();
+}
+
+bool LoopMetadata::assignLoopMetadata(Loop &L) {
+  this->L = &L;
+  Context = &L.getHeader()->getParent()->getContext();
+
+  std::optional<int> MinIterCount = getMinTripCount(&L);
+
+  // dump loop summary
+  LLVM_DEBUG(if (L.getLoopPreheader()) {
+    dbgs() << "Preheader:" << L.getLoopPreheader()->getName() << "\n";
+  });
+  LLVM_DEBUG(dbgs() << "Header:" << L.getHeader()->getName() << "\n");
+
+  if (MinIterCount.has_value() && MinIterCount.value() > 0) {
+    this->MinIterCount = MinIterCount.value();
+
+    LLVM_DEBUG(L.getHeader()->getParent()->dump(););
+    addAssumeToLoopHeader();
+    return true;
+  }
+
+  return false;
+}
+bool LoopMetadata::canExtractIncrement(const SCEV *S) {
+  const SCEVAddRecExpr *AR = dyn_cast<SCEVAddRecExpr>(S);
+  if (!AR)
+    return false;
+
+  const SCEVConstant *SCEVConst =
+      dyn_cast<SCEVConstant>(AR->getStepRecurrence(*SE));
+  if (!SCEVConst) {
+    return false;
+  }
+  LoopStepSize = getSCEVStepSize(SCEVConst);
+  IsLoopIncrementing = LoopStepSize > 0;
+  return true;
+}
+
+Value *LoopMetadata::calcMinIterValue(const SCEV *S, int MinIterCount,
+                                      LLVMContext *Context) {
+  assert(MinIterCount > 0);
+
+  // extract loop counter increment/decrement
+  const SCEVAddRecExpr *AR = dyn_cast<SCEVAddRecExpr>(S);
+  if (!AR) {
+    LLVM_DEBUG(dbgs() << "LoopMetadata-Warning: SCEV S i not a SCEVAddRecExpr ";
+               S->dump());
+    return nullptr;
+  }
+  const SCEV *ConstExpr = AR->getOperand(1);
+  if (!AR->isAffine() || !isa<SCEVConstant>(ConstExpr)) {
+    LLVM_DEBUG(dbgs() << "LoopMetadata-Warning: Unknown SCEVAddRecExpr ";
+               AR->dump());
+    return nullptr;
+  }
+
+  int MinIterValue = std::abs(LoopStepSize * MinIterCount);
+  // calculate the minimum iteration value, since SGE is used, subtract 1
+  MinIterValue--;
+
+  return llvm::ConstantInt::get(llvm::Type::getInt32Ty(*Context), MinIterValue,
+                                true);
+}
+
+Value *getLoopInvariant(Value *Op0, Value *Op1, ScalarEvolution *SE) {
+  if ((SE->isSCEVable(Op0->getType()) &&
+       (SCEVAddExpr::classof(SE->getSCEV(Op0)) ||
+        SCEVAddRecExpr::classof(SE->getSCEV(Op0)))) ||
+      isa<PHINode>(Op0))
+    return Op1;
```

Comment:
It is not trivial to assume that Op1 is loopinvariant if Op0 satisfies nice loop-affine properties. I think the function should at least mention the preconditions for Op0 and Op1

Reply:
added a `bool Loop::isLoopInvariant(const Value *V) const` check, so that I get the valid loop invariant operand.

Reply:
I rewrote this to check the individual operand before returning it and added a PHI check and a dominance check (that the operand is in the loop header or outside of the loop). This dominance check is relevant to detect LICM that the previous pass generated.